### ISSUE TITLE
firebaseStorageDownloadTokens from object metadata

### ISSUE THREAD

**Issue by envyram** [firebase/extensions#34]
### [READ] Step 1: Are you in the right place?

 Yes

### [REQUIRED] Step 2: Extension name

This feature request is for extension: **\_** (`storage-resize-images`)

### What feature would you like to see?

Any object on firebase storage contains firebaseStorageDownloadTokens from object metadata. However this is not available for resized images

### How would you use it?

I use firebaseStorageDownloadTokens to construct the downloadUrl for the object

**Comment by rosariopfernandes** [comment]
@envyram I'm not sure what you mean. Looking at the [StorageMetadata reference](https://firebase.google.com/docs/reference/android/com/google/firebase/storage/StorageMetadata) I don't see any downloadTokens. Can you tell us how do you usually get this metadata? Is it custom metadata?

**Comment by envyram** [comment]
@rosariopfernandes  

In firebase cloud functions:

exports.thumbnailHandler = functions.storage.object().onFinalize(async (object) => {
   console.log(object.metadata.firebaseStorageDownloadTokens);
}

This is available for all of my other objects in the bucket. But is not available only for the thumbnails created from the extension

**Comment by rosariopfernandes** [comment]
@envyram Thanks for the useful code snippet. I couldn't find this firebaseStorageDownloadTokens anywhere on the docs, and [this post](https://stackoverflow.com/a/56010225/5861618) actually explains why:

> Doug Stevenson pointed out that firebaseStorageDownloadTokens is not an official Google Cloud Storage feature. You won't find it in any Google documentation, and there's no promise it will be in future version of @google-cloud. I like firebaseStorageDownloadTokens because it's the only way to get what I want, but it has a "smell" that it's not safe to use.

Since this is not official, I'm not sure if we should have it on Firebase Extensions, but I'll let the Firebase Team decide.

**Comment by envyram** [comment]
@rosariopfernandes that's a great discussion. "getDownloadURL() from Node" is what I am looking for actually. Without that the extension on Firebase is not quite useful

**Comment by jhuleatt** [comment]
I chatted with the Cloud Storage for Firebase team about `firebaseStorageDownloadTokens`. The field should be unique per object. If it isn't set on an object, it will be set automatically by the Cloud Storage for Firebase api when `getDownloadURL()` is called.

TL;DR we shouldn't copy it over in the extension.

**Comment by rosariopfernandes** [comment]
@jhuleatt thanks!

@envyram have you tried using the [Google Cloud Client Library](https://cloud.google.com/storage/docs/reference/libraries) for NodeJS? The docs contain a snippet on how to [download a file](https://cloud.google.com/storage/docs/downloading-objects#storage-download-object-nodejs):
```js
await storage
  .bucket(bucketName)
  .file(srcFilename)
  .download({destination: 'example.jpg'});

console.log(
  `gs://${bucketName}/${srcFilename} downloaded to ${destFilename}.`
);
```

**Comment by Ehesp** [comment]
XFN Team: Couple of options available here; the extension could initialise `getDownloadURL()` or leave this to the user (maybe improved documentation is needed?).

**Comment by jhuleatt** [comment]
Maybe the extension can just call `getDownloadURL` after it creates the new image, but not block on that call? That way, by the time a client tries to access the image, the storage download tokens should be created.

Note: The missing `firebaseStorageDownloadTokens` seem to break the image previews in the Firebase console too

**Comment by lsps9150414** [comment]
> Maybe the extension can just call getDownloadURL after it creates the new image, but not block on that call? That way, by the time a client tries to access the image, the storage download tokens should be created.
> 
> Note: The missing firebaseStorageDownloadTokens seem to break the image previews in the Firebase console too

Hey @jhuleatt thanks for the summary. So is the current plan to call `getDownloadURL` by the extension so that we can get `firebaseStorageDownloadTokens` in the metadata via `storage.object().onFinalize()` triggers?

**Comment by Ehesp** [comment]
There is overlap here with issue https://github.com/firebase/extensions/issues/140 - the issue is that the Admin SDKs (in our case NodeJS) cannot call `getDownloadURL`, as this needs to come from a client instance. We're currently discussing the best solution/idea to handle the tokens/download URLs.

**Comment by dfdgsdfg** [comment]
Also, `adminStorage.bucket(bucketName).file(filePath).getSignedUrl(options)` create signedUrl but it is not works.

```xml
<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>
The request signature we calculated does not match the signature you provided. Check your Google secret key and signing method.
</Message>
<StringToSign>
GET 1895632947 /project.appspot.com/users/NK8bDKYlaofgwkX6cxy14tqpsGa2/room_backgrounds/room_bg_1584592936559_Ib6R9sawuIQ_200x200.jpeg
</StringToSign>
</Error>
```

If I create access token on web ui after created signedUrl, it still does not work.

**Comment by laurenzlong** [comment]
This appears to be a duplicate of https://github.com/firebase/extensions/issues/140, which will be solved by https://github.com/firebase/extensions/pull/279
